### PR TITLE
Deinterleave storage

### PR DESCRIPTION
**Pull request by dsharletg**
This PR has a number of improvements to the eliminate interleaves optimization in Hexagon:

- The old optimization was a bit too conservative about when it removed interleaves, it would fail to see opportunities to remove interleaves through Let and LetStmt.
- Simplified the logic around dealing with boolean/mask vectors.
- The optimization now attempts to deinterleave buffers which are only stored interleaved and deinterleaved upon being loaded from. This means that reductions which widen, followed by usages that narrow the result, will be deinterleaved. 
- To make it easier to implement, moved some crufty logic to implement trunc_satuh(vw) to the runtime.

This significantly impacts some workloads, the widening reduction test drops in simulated cycles by ~1/3.

### CODE REVIEW

**Comment by ronlieb** [comment]
LGTM, although i did not see any performance change on the CONV3x3a16 test case i have that contains RDoms, i can remeasure after its merged

**Comment by dsharletg** [comment]
Can you share the code for that benchmark?

**Comment by ronlieb** [comment]
Sending it separately …

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -850,10 +863,20 @@ class EliminateInterleaves : public IRMutator {
     template <typename NodeType, typename LetType>
     void visit_let(NodeType &result, const LetType *op) {
         Expr value = mutate(op->value);
-        string deinterleaved_name = op->name + ".deinterleaved";
+        string deinterleaved_name;
         NodeType body;
-        if (is_native_interleave(value)) {
+        // Other code in this mutator needs to be able to tell the
+        // difference between a Let that yields a deinterleave, and a
+        // let that has a remoable deinterleave.
```
Comment (abadams) [inline src/HexagonOptimize.cpp]: remoable

File: src/HexagonOptimize.cpp [hunk 2]
```diff
@@ -754,8 +760,20 @@ class EliminateInterleaves : public IRMutator {
...
         if (var) {
-            internal_assert(vars.contains(var->name + ".deinterleaved"));
-            return Variable::make(var->type, var->name + ".deinterleaved");
+            if (vars.contains(var->name + ".deinterleaved")) {
+                return Variable::make(var->type, var->name + ".deinterleaved");
+            } else if (vars.contains(var->name + ".weak_deinterleaved")) {
+                return Variable::make(var->type, var->name + ".weak_deinterleaved");
+            }
+        }
+        const Let *let = x.as<Let>();
```
Comment (abadams) [inline src/HexagonOptimize.cpp]: Declaration could be folded into the if condition

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -683,59 +683,65 @@ class OptimizePatterns : public IRMutator {
...
-        ++allow_interleave_expr;
-        return ret;
+    // Check if x is an expression that is either an interleave, or
+    // transitively is an interleave.
+    bool yields_removable_interleave(Expr x) {
+        if (is_native_interleave(x)) {
+            return true;
+        }
+
+        const Let *let = x.as<Let>();
```
Comment (abadams) [inline src/HexagonOptimize.cpp]: Declaration could be folded into the if condition

**Comment by abadams** [comment]
Could use more comments explaining what weak_deinterleaved means

**Comment by abadams** [comment]
Seems like the new test fails for opencl: 

https://buildbot.halide-lang.org/master/builders/linux-64-gcc49-381/builds/226

**Pull request merged**